### PR TITLE
rockchip: optimize the link mechanism for SRAM code

### PR DESCRIPTION
**Pull request by rkchrome**
Add the common extra.ld.S and customized rk3399.ld.S to extend
to more features for different platforms.
For example, we can add SRAM section and specific address to
load there if we need it, and the common bl31.ld.S not need to
be modified.

Therefore, we can remove the unused codes which copying explicitly
from the function pmusram_prepare(). It looks like more clear.

Change-Id: Ibffa2da5e8e3d1d2fca80085ebb296ceb967fce8
(cherry-picked pick from Xing Zheng <zhengxing@rock-chips.com> works
 on https://chromium-review.googlesource.com/#/c/389114/)

### CODE REVIEW

**Comment by davidcunado-arm** [comment]
jenkins: test this please

File: plat/rockchip/rk3399/drivers/pmu/rk3399m0.h [hunk 1]
```diff
@@ -0,0 +1,10 @@
+#ifndef __RK3399M0_H__
```
Comment (davidcunado-arm) [inline plat/rockchip/rk3399/drivers/pmu/rk3399m0.h]: Needs copyright boiler-plate info here.
This would need to be addressed ahead of a merge.

**Comment by davidcunado-arm** [comment]
Also, build failed as follows (debug shown here, but release also failed).

./build/rk3368/debug/bl31/pmu_sram.o: In function `rockchip_plat_sram_mmu_el3':
/work/workspace/trusted-firmware-build/BUILD_CONFIG/rk3368-default/MODE/Debug/label/platforms/plat/rockchip/common/pmusram/pmu_sram.c:50: undefined reference to `__bl31_sram_text_start'
/work/workspace/trusted-firmware-build/BUILD_CONFIG/rk3368-default/MODE/Debug/label/platforms/plat/rockchip/common/pmusram/pmu_sram.c:50: undefined reference to `__bl31_sram_text_end'
/work/workspace/trusted-firmware-build/BUILD_CONFIG/rk3368-default/MODE/Debug/label/platforms/plat/rockchip/common/pmusram/pmu_sram.c:50: undefined reference to `__bl31_sram_text_start'
/work/workspace/trusted-firmware-build/BUILD_CONFIG/rk3368-default/MODE/Debug/label/platforms/plat/rockchip/common/pmusram/pmu_sram.c:50: undefined reference to `__bl31_sram_text_end'
/work/workspace/trusted-firmware-build/BUILD_CONFIG/rk3368-default/MODE/Debug/label/platforms/plat/rockchip/common/pmusram/pmu_sram.c:57: undefined reference to `__bl31_sram_data_start'
/work/workspace/trusted-firmware-build/BUILD_CONFIG/rk3368-default/MODE/Debug/label/platforms/plat/rockchip/common/pmusram/pmu_sram.c:57: undefined reference to `__bl31_sram_data_end'
/work/workspace/trusted-firmware-build/BUILD_CONFIG/rk3368-default/MODE/Debug/label/platforms/plat/rockchip/common/pmusram/pmu_sram.c:57: undefined reference to `__bl31_sram_data_start'
/work/workspace/trusted-firmware-build/BUILD_CONFIG/rk3368-default/MODE/Debug/label/platforms/plat/rockchip/common/pmusram/pmu_sram.c:57: undefined reference to `__bl31_sram_data_end'
make: *** [build/rk3368/debug/bl31/bl31.elf] Error 1

**Comment by davidcunado-arm** [comment]
Also, build failed as follows (debug shown here, but release also failed).

./build/rk3368/debug/bl31/pmu_sram.o: In function `rockchip_plat_sram_mmu_el3':
/work/workspace/trusted-firmware-build/BUILD_CONFIG/rk3368-default/MODE/Debug/label/platforms/plat/rockchip/common/pmusram/pmu_sram.c:50: undefined reference to `__bl31_sram_text_start'
/work/workspace/trusted-firmware-build/BUILD_CONFIG/rk3368-default/MODE/Debug/label/platforms/plat/rockchip/common/pmusram/pmu_sram.c:50: undefined reference to `__bl31_sram_text_end'
/work/workspace/trusted-firmware-build/BUILD_CONFIG/rk3368-default/MODE/Debug/label/platforms/plat/rockchip/common/pmusram/pmu_sram.c:50: undefined reference to `__bl31_sram_text_start'
/work/workspace/trusted-firmware-build/BUILD_CONFIG/rk3368-default/MODE/Debug/label/platforms/plat/rockchip/common/pmusram/pmu_sram.c:50: undefined reference to `__bl31_sram_text_end'
/work/workspace/trusted-firmware-build/BUILD_CONFIG/rk3368-default/MODE/Debug/label/platforms/plat/rockchip/common/pmusram/pmu_sram.c:57: undefined reference to `__bl31_sram_data_start'
/work/workspace/trusted-firmware-build/BUILD_CONFIG/rk3368-default/MODE/Debug/label/platforms/plat/rockchip/common/pmusram/pmu_sram.c:57: undefined reference to `__bl31_sram_data_end'
/work/workspace/trusted-firmware-build/BUILD_CONFIG/rk3368-default/MODE/Debug/label/platforms/plat/rockchip/common/pmusram/pmu_sram.c:57: undefined reference to `__bl31_sram_data_start'
/work/workspace/trusted-firmware-build/BUILD_CONFIG/rk3368-default/MODE/Debug/label/platforms/plat/rockchip/common/pmusram/pmu_sram.c:57: undefined reference to `__bl31_sram_data_end'
make: *** [build/rk3368/debug/bl31/bl31.elf] Error 1

File: plat/rockchip/common/pmusram/pmu_sram.c [hunk 2]
```diff
@@ -36,3 +39,30 @@ void u32_align_cpy(uint32_t *dst, const uint32_t *src, size_t bytes)
...
+	mmap_add_region((unsigned long)&__bl31_sram_data_start,
+			(unsigned long)&__bl31_sram_data_start,
+			sram_size, MT_MEMORY | MT_RW | MT_SECURE);
+}
+
+void plat_rockchip_mem_prepare(void)
+{	/*
+	 * The code for resuming cpu from suspend must be excuted in pmusram.
+	 * Copy the code into pmusram.
+	 */
```
Comment (Caesar-github) [inline plat/rockchip/common/pmusram/pmu_sram.c]: Remove it.

**Comment by Caesar-github** [comment]
Sorry, without thinking about the RK3368 part. Can you try it again?

**Comment by davidcunado-arm** [comment]
Hi!
Missed an issue in the commit messages, But otherwise the build issues are now resoved.
Thanks!

**Comment by davidcunado-arm** [comment]
Just realised that you may not be aware of recent change to the contribution policy that we announed with v1.3 (see [https://github.com/ARM-software/tf-issues/issues/429]).
Also, [PR722](https://github.com/ARM-software/arm-trusted-firmware/pull/722) contains the changes to the contributing document and explains the change in more detail.

**Comment by rkchrome** [comment]
Yup, updated now.

**Comment by davidcunado-arm** [comment]
jenkins: test this please

**Pull request merged**